File: backend/src/Modules/Catalog/Catalog.Domain/Variants/Variant.cs
using Catalog.Domain.Variants.Events;
using SharedKernel;

namespace Catalog.Domain.Variants;

/// <summary>
/// Ürün varyasyon eksenlerini (renk, beden vb.) tanımlayan kök varlık.
/// SKU kombinasyonunu değil; eksen adını, seçim stilini, sıralamasını ve seçilebilir değerleri yönetir.
/// </summary>
/// <example>
/// "Renk" varyantı key değeri RENK, SelectionStyle Color, SortOrder 1, Slicer true ile oluşturulur;
/// "Kırmızı" (#FF0000) ve "Mavi" (#0000FF) değerleri eklenir. "Beden" varyantı List stiliyle ayrı tanımlanır.
/// </example>
public sealed class Variant : AggregateRoot<Guid>
{
    private readonly List<VariantValue> _values = [];

    private Variant()
    {
    }

    private Variant(Guid id, VariantKey key, string name, SelectionStyle selectionStyle, int sortOrder, bool slicer)
        : base(id)
    {
        Key = key;
        Name = name;
        SelectionStyle = selectionStyle;
        SortOrder = sortOrder;
        Slicer = slicer;
    }

    /// <summary>Gets varyant türünü benzersiz tanımlayan anahtar; oluşturulurken adından türetilir.</summary>
    /// <example>RENK.</example>
    public VariantKey Key { get; private set; } = null!;

    /// <summary>Gets varyant türünün adı.</summary>
    public string Name { get; private set; } = string.Empty;

    /// <summary>Gets kullanıcı arayüzünde seçim stili.</summary>
    public SelectionStyle SelectionStyle { get; private set; }

    /// <summary>Gets ürün içindeki görüntüleme sırası.</summary>
    public int SortOrder { get; private set; }

    /// <summary>Gets a value indicating whether filtreleme (slicer) olarak kullanılıp kullanılmadığı.</summary>
    public bool Slicer { get; private set; }

    /// <summary>Gets türe ait seçilebilir değerler.</summary>
    public IReadOnlyCollection<VariantValue> Values => _values.AsReadOnly();

    /// <summary>Yeni varyant türü oluşturur ve <see cref="VariantCreated"/> alan olayını yayımlar.</summary>
    /// <param name="name">Varyant türü adı.</param>
    /// <param name="selectionStyle">Kullanıcı arayüzü seçim stili.</param>
    /// <param name="sortOrder">Görüntüleme sırası.</param>
    /// <param name="slicer">Filtreleme (slicer) olarak kullanılsın mı.</param>
    /// <param name="key">Açık anahtar; boş bırakılırsa adından türetilir.</param>
    public static Result<Variant> Create(
        string name,
        SelectionStyle selectionStyle,
        int sortOrder,
        bool slicer = false,
        string? key = null)
    {
        if (string.IsNullOrWhiteSpace(name))
        {
            return Result.Failure<Variant>(Error.Validation("Variant type name is required."));
        }

        var trimmedName = name.Trim();
        var keyResult = VariantKey.FromOptional(key, trimmedName);
        if (keyResult.IsFailure)
        {
            return Result.Failure<Variant>(keyResult.Error);
        }

        var variant = new Variant(
            Guid.NewGuid(),
            keyResult.Value,
            trimmedName,
            selectionStyle,
            sortOrder,
            slicer);

        variant.RaiseDomainEvent(new VariantCreated(variant.Id, variant.Key.Value));
        return Result.Success(variant);
    }

    /// <summary>Varyant türünün adını, seçim stilini, sırasını ve slicer bayrağını günceller.</summary>
    /// <param name="name">Yeni tür adı.</param>
    /// <param name="selectionStyle">Yeni seçim stili.</param>
    /// <param name="sortOrder">Yeni görüntüleme sırası.</param>
    /// <param name="slicer">Filtreleme (slicer) olarak kullanılsın mı.</param>
    public Result Rename(string name, SelectionStyle selectionStyle, int sortOrder, bool slicer)
    {
        if (string.IsNullOrWhiteSpace(name))
        {
            return Result.Failure(Error.Validation("Variant type name is required."));
        }

        Name = name.Trim();
        SelectionStyle = selectionStyle;
        SortOrder = sortOrder;
        Slicer = slicer;
        return Result.Success();
    }

    /// <summary>Varyant türüne yeni seçilebilir değer ekler.</summary>
    /// <param name="label">Değer etiketi.</param>
    /// <param name="color">Görsel renk kodu; opsiyonel.</param>
    /// <param name="imageUrl">Görsel URL'si; opsiyonel.</param>
    /// <param name="key">Değer anahtarı; boş bırakılırsa etiketten türetilir.</param>
    /// <param name="sortOrder">Tür içindeki görüntüleme sırası.</param>
    public Result<VariantValue> AddValue(
        string label,
        string? color,
        string? imageUrl,
        string? key,
        int sortOrder)
    {
        if (string.IsNullOrWhiteSpace(label))
        {
            return Result.Failure<VariantValue>(Error.Validation("Variant value label is required."));
        }

        var trimmedLabel = label.Trim();
        if (_values.Any(v => string.Equals(v.Label, trimmedLabel, StringComparison.OrdinalIgnoreCase)))
        {
            return Result.Failure<VariantValue>(
                Error.Conflict("Variant value label must be unique within the type."));
        }

        var keyResult = VariantKey.FromOptional(key, trimmedLabel);
        if (keyResult.IsFailure)
        {
            return Result.Failure<VariantValue>(keyResult.Error);
        }

        if (_values.Any(v => string.Equals(v.Key.Value, keyResult.Value.Value, StringComparison.OrdinalIgnoreCase)))
        {
            return Result.Failure<VariantValue>(
                Error.Conflict("Variant value key must be unique within the type."));
        }

        var value = new VariantValue(
            Guid.NewGuid(),
            keyResult.Value,
            trimmedLabel,
            string.IsNullOrWhiteSpace(color) ? null : color.Trim(),
            string.IsNullOrWhiteSpace(imageUrl) ? null : imageUrl.Trim(),
            sortOrder);

        _values.Add(value);
        return Result.Success(value);
    }

    /// <summary>Mevcut varyant değerinin etiket, görsel ve anahtar bilgilerini günceller.</summary>
    /// <param name="valueId">Güncellenecek değer tanımlayıcısı.</param>
    /// <param name="label">Yeni değer etiketi.</param>
    /// <param name="color">Yeni renk kodu; opsiyonel.</param>
    /// <param name="imageUrl">Yeni görsel URL'si; opsiyonel.</param>
    /// <param name="key">Yeni değer anahtarı; boş bırakılırsa etiketten türetilir.</param>
    /// <param name="sortOrder">Yeni görüntüleme sırası.</param>
    public Result UpdateValue(
        Guid valueId,
        string label,
        string? color,
        string? imageUrl,
        string? key,
        int sortOrder)
    {
        if (string.IsNullOrWhiteSpace(label))
        {
            return Result.Failure(Error.Validation("Variant value label is required."));
        }

        var value = _values.FirstOrDefault(v => v.Id == valueId);
        if (value is null)
        {
            return Result.Failure(Error.NotFound("Variant value not found."));
        }

        var trimmedLabel = label.Trim();
        if (_values.Any(v => v.Id != valueId &&
                             string.Equals(v.Label, trimmedLabel, StringComparison.OrdinalIgnoreCase)))
        {
            return Result.Failure(Error.Conflict("Variant value label must be unique within the type."));
        }

        var keyResult = VariantKey.FromOptional(key, trimmedLabel);
        if (keyResult.IsFailure)
        {
            return Result.Failure(keyResult.Error);
        }

        if (_values.Any(v => v.Id != valueId &&
                             string.Equals(v.Key.Value, keyResult.Value.Value, StringComparison.OrdinalIgnoreCase)))
        {
            return Result.Failure(Error.Conflict("Variant value key must be unique within the type."));
        }

        value.Update(
            keyResult.Value,
            trimmedLabel,
            string.IsNullOrWhiteSpace(color) ? null : color.Trim(),
            string.IsNullOrWhiteSpace(imageUrl) ? null : imageUrl.Trim(),
            sortOrder);

        return Result.Success();
    }

    /// <summary>Varyant türünden bir değeri kaldırır.</summary>
    /// <param name="valueId">Kaldırılacak değer tanımlayıcısı.</param>
    public Result RemoveValue(Guid valueId)
    {
        var value = _values.FirstOrDefault(v => v.Id == valueId);
        if (value is null)
        {
            return Result.Failure(Error.NotFound("Variant value not found."));
        }

        _values.Remove(value);
        return Result.Success();
    }

    /// <summary>Kalıcı depodan yüklenen değerleri aggregate'e bağlar.</summary>
    /// <param name="values">Türe ait değer koleksiyonu.</param>
    internal void LoadValues(IEnumerable<VariantValue> values)
    {
        _values.Clear();
        _values.AddRange(values);
    }
}

File: backend-go/internal/modules/catalog/domain/variants/variant.go
// Package variants, ürün varyasyon eksenlerini (renk, beden vb.) tanımlayan
// kök varlığı içerir (.NET Catalog.Domain.Variants karşılığı). SKU
// kombinasyonunu değil; eksen adını, seçim stilini, sıralamasını ve
// seçilebilir değerleri yönetir.
package variants

import (
	"strings"

	"github.com/google/uuid"

	"pimly.commerslab/backend-go/internal/modules/catalog/domain/keygen"
	"pimly.commerslab/backend-go/internal/sharedkernel"
)

// SelectionStyle, varyant değerlerinin kullanıcı arayüzünde nasıl sunulacağını
// belirler; kabloda ve veritabanında küçük harfli dizgi taşınır ("list"/"color").
type SelectionStyle string

// Seçim stilleri.
const (
	// StyleList: değerler düz liste olarak sunulur.
	StyleList SelectionStyle = "list"

	// StyleColor: değerler renk yuvarlakları olarak sunulur.
	StyleColor SelectionStyle = "color"
)

// ParseSelectionStyle, kullanıcı girdisini stile çözer (.NET Enum.Parse
// ignoreCase karşılığı); tanınmayan değer için ok=false döner.
func ParseSelectionStyle(value string) (SelectionStyle, bool) {
	switch strings.ToLower(strings.TrimSpace(value)) {
	case string(StyleList):
		return StyleList, true
	case string(StyleColor):
		return StyleColor, true
	default:
		return "", false
	}
}

// Value, bir varyant türüne ait seçilebilir değerdir; etiket, opsiyonel renk
// ve görsel bilgileriyle (.NET VariantValue karşılığı).
type Value struct {
	// ID, değerin benzersiz kimliğidir.
	ID uuid.UUID

	// Key, değeri benzersiz tanımlayan anahtardır; verilmezse etiketten türetilir.
	Key string

	// Label, değerin görünen etiketidir (tür içinde büyük/küçük harf duyarsız benzersiz).
	Label string

	// Color, görsel gösterim için renk kodudur; opsiyonel.
	Color *string

	// ImageURL, görsel gösterim için resim adresidir; opsiyonel.
	ImageURL *string

	// SortOrder, tür içindeki görüntüleme sırasıdır.
	SortOrder int
}

// Variant, varyant türünün kök varlığıdır.
type Variant struct {
	// ID, türün benzersiz kimliğidir.
	ID uuid.UUID

	// Key, türü benzersiz tanımlayan anahtardır (RENK); verilmezse adından türetilir.
	Key string

	// Name, türün görünen adıdır (Renk).
	Name string

	// SelectionStyle, kullanıcı arayüzü seçim stilidir.
	SelectionStyle SelectionStyle

	// SortOrder, ürün içindeki görüntüleme sırasıdır.
	SortOrder int

	// Slicer, türün filtreleme (slicer) ekseni olup olmadığını belirtir;
	// tenant başına en fazla bir tür slicer olabilir.
	Slicer bool

	// Values, türe ait seçilebilir değerlerdir.
	Values []*Value
}

// keyFromOptional, açık anahtar verilmişse doğrular, verilmemişse addan türetir
// (.NET VariantKey.FromOptional karşılığı).
func keyFromOptional(key *string, fallbackName string) sharedkernel.ResultOf[string] {
	if key == nil || strings.TrimSpace(*key) == "" {
		return keygen.FromName(fallbackName)
	}
	return keygen.ValidateExplicit(*key)
}

// NewVariant, doğrulanmış yeni bir varyant türü oluşturur.
func NewVariant(name string, style SelectionStyle, sortOrder int, slicer bool, key *string) sharedkernel.ResultOf[*Variant] {
	if strings.TrimSpace(name) == "" {
		return sharedkernel.FailOf[*Variant](sharedkernel.NewValidationError("Variant type name is required."))
	}
	trimmed := strings.TrimSpace(name)
	keyResult := keyFromOptional(key, trimmed)
	if keyResult.IsFailure() {
		return sharedkernel.FailOf[*Variant](keyResult.Err())
	}
	return sharedkernel.OkOf(&Variant{
		ID:             uuid.New(),
		Key:            keyResult.Value(),
		Name:           trimmed,
		SelectionStyle: style,
		SortOrder:      sortOrder,
		Slicer:         slicer,
	})
}

// Rename, türün adını, seçim stilini, sırasını ve slicer bayrağını günceller
// (anahtar değişmez).
func (v *Variant) Rename(name string, style SelectionStyle, sortOrder int, slicer bool) sharedkernel.Result {
	if strings.TrimSpace(name) == "" {
		return sharedkernel.Fail(sharedkernel.NewValidationError("Variant type name is required."))
	}
	v.Name = strings.TrimSpace(name)
	v.SelectionStyle = style
	v.SortOrder = sortOrder
	v.Slicer = slicer
	return sharedkernel.Ok()
}

// AddValue, türe yeni değer ekler; etiket ve anahtar tür içinde büyük/küçük
// harf duyarsız benzersiz olmalıdır.
func (v *Variant) AddValue(label string, color, imageURL, key *string, sortOrder int) sharedkernel.ResultOf[*Value] {
	if strings.TrimSpace(label) == "" {
		return sharedkernel.FailOf[*Value](sharedkernel.NewValidationError("Variant value label is required."))
	}
	trimmedLabel := strings.TrimSpace(label)
	for _, existing := range v.Values {
		if strings.EqualFold(existing.Label, trimmedLabel) {
			return sharedkernel.FailOf[*Value](
				sharedkernel.NewConflictError("Variant value label must be unique within the type."))
		}
	}

	keyResult := keyFromOptional(key, trimmedLabel)
	if keyResult.IsFailure() {
		return sharedkernel.FailOf[*Value](keyResult.Err())
	}
	for _, existing := range v.Values {
		if strings.EqualFold(existing.Key, keyResult.Value()) {
			return sharedkernel.FailOf[*Value](
				sharedkernel.NewConflictError("Variant value key must be unique within the type."))
		}
	}

	value := &Value{
		ID:        uuid.New(),
		Key:       keyResult.Value(),
		Label:     trimmedLabel,
		Color:     normalizeOptional(color),
		ImageURL:  normalizeOptional(imageURL),
		SortOrder: sortOrder,
	}
	v.Values = append(v.Values, value)
	return sharedkernel.OkOf(value)
}

// UpdateValue, mevcut değerin etiket/görsel/anahtar/sıra bilgilerini günceller;
// benzersizlik kuralları korunur.
func (v *Variant) UpdateValue(valueID uuid.UUID, label string, color, imageURL, key *string, sortOrder int) sharedkernel.Result {
	if strings.TrimSpace(label) == "" {
		return sharedkernel.Fail(sharedkernel.NewValidationError("Variant value label is required."))
	}
	value := v.findValue(valueID)
	if value == nil {
		return sharedkernel.Fail(sharedkernel.NewNotFoundError("Variant value not found."))
	}

	trimmedLabel := strings.TrimSpace(label)
	for _, existing := range v.Values {
		if existing.ID != valueID && strings.EqualFold(existing.Label, trimmedLabel) {
			return sharedkernel.Fail(sharedkernel.NewConflictError("Variant value label must be unique within the type."))
		}
	}
	keyResult := keyFromOptional(key, trimmedLabel)
	if keyResult.IsFailure() {
		return sharedkernel.Fail(keyResult.Err())
	}
	for _, existing := range v.Values {
		if existing.ID != valueID && strings.EqualFold(existing.Key, keyResult.Value()) {
			return sharedkernel.Fail(sharedkernel.NewConflictError("Variant value key must be unique within the type."))
		}
	}

	value.Key = keyResult.Value()
	value.Label = trimmedLabel
	value.Color = normalizeOptional(color)
	value.ImageURL = normalizeOptional(imageURL)
	value.SortOrder = sortOrder
	return sharedkernel.Ok()
}

// RemoveValue, türden bir değeri kaldırır.
func (v *Variant) RemoveValue(valueID uuid.UUID) sharedkernel.Result {
	for i, existing := range v.Values {
		if existing.ID == valueID {
			v.Values = append(v.Values[:i], v.Values[i+1:]...)
			return sharedkernel.Ok()
		}
	}
	return sharedkernel.Fail(sharedkernel.NewNotFoundError("Variant value not found."))
}

// findValue, kimlikle değeri döner; yoksa nil.
func (v *Variant) findValue(id uuid.UUID) *Value {
	for _, existing := range v.Values {
		if existing.ID == id {
			return existing
		}
	}
	return nil
}

// normalizeOptional, opsiyonel dizgiyi kırpar; boş değer nil'e çevrilir.
func normalizeOptional(s *string) *string {
	if s == nil {
		return nil
	}
	trimmed := strings.TrimSpace(*s)
	if trimmed == "" {
		return nil
	}
	return &trimmed
}
